tests/adjoint_2d_cylindrical/inverse.py

```diff
@@ -0,0 +1,408 @@
+
+    # Without a restart to continue from, our initial guess is the final state of the forward run
+    # We need to project the state from Q2 into Q1
+    Tic = Function(Q1, name="Initial_Temperature")
+    T_0 = Function(DQ, name="T_0")  # Temperature for zeroth time-step
+    Taverage = Function(Q1, name="Average_Temperature")
+
+    # Initialise the control to final state temperature from forward model.
+    checkpoint_file = CheckpointFile("Checkpoint_State.h5", "r")
+    Tic.project(
+        checkpoint_file.load_function(mesh, "Temperature", idx=max_timesteps - 1)
+    )
+    Taverage.project(checkpoint_file.load_function(mesh, "Average_Temperature", idx=0))
+
+    # Temperature function in Q2, where we solve the equations
+    T = Function(DQ, name="Temperature")
+
+    mu = get_viscosity(r, T, u)
+
+    # Configure approximation
+    approximation = BoussinesqApproximation(Ra, mu=mu)
+
+    # Nullspaces and near-nullspaces:
+    Z_nullspace = create_stokes_nullspace(Z, closed=True, rotational=not uimposed)
+    Z_near_nullspace = create_stokes_nullspace(
+        Z, closed=False, rotational=True, translations=[0, 1]
+    )
+
+    # Generate a surface velocity reference
+    uobs = Function(V, name="uobs")
+
+    # Free-slip velocity boundary condition on all sides
+    stokes_bcs = {
+        "bottom": {"un": 0},
+        "top": {"u": uobs} if uimposed else {"un": 0},  # surface velocity is free-slip unless imposed
+    }
+    temp_bcs = {
+        "bottom": {"T": 1.0},
+        "top": {"T": 0.0},
+    }
+
+    energy_solver = EnergySolver(
+        T, u, approximation, delta_t, ImplicitMidpoint, bcs=temp_bcs
+    )
+
+    stokes_solver = StokesSolver(
+        z,
+        T,
+        approximation,
+        bcs=stokes_bcs,
+        nullspace=Z_nullspace,
+        transpose_nullspace=Z_nullspace,
+        near_nullspace=Z_near_nullspace,
+        solver_parameters="direct"
+    )
+
+    # Control variable for optimisation
+    control = Control(Tic)
+
+    # If we are using surface veolocit misfit in the functional
+    if alpha_u > 0:
+        u_misfit = 0.0
+
+    # We need to project the initial condition from Q1 to Q2,
+    # and impose the boundary conditions at the same time
+    T_0.project(Tic, bcs=energy_solver.strong_bcs)
+    T.assign(T_0)
+
+    # if the weighting for misfit terms non-positive, then no need to integrate in time
+    min_timesteps = 0 if any([w > 0 for w in [alpha_T, alpha_u]]) else max_timesteps
+
+    # Populate the tape by running the forward simulation
+    for timestep in tape.timestepper(iter(range(min_timesteps, max_timesteps))):
+        if alpha_u > 0 or uimposed:
+            # Update the accumulated surface velocity misfit using the observed value
+            uobs.assign(checkpoint_file.load_function(mesh, name="Velocity", idx=timestep))
+
+        stokes_solver.solve()
+        energy_solver.solve()
+
+        if alpha_u > 0:
```

Comment:
Just for my understanding: I thought the tape ignores `if` conditions, so how do the ones included here work?

Reply:
I think tape actually ignores firedrake `conditional` constructs. With a regular Python `if`, something will either be on the tape or not. That is, the condition isn't dynamically re-evaluated when the tape is replayed. But for static coefficients like this, that's not an issue.

Reply:
Thank you!